### ISSUE TITLE
[iOS] "getLocation" not resolving

### ISSUE THREAD

**Issue by maerlynflagg** [Lyokone/flutterlocation#621]
**Describe the bug**
a show a waiter around the function `getLocation()`. so a customer found out, the waiter is not hiding. the reason is, that the line `await _locationPlugin.getLocation()` is not resolving.

after testing and debugging i found out, that the function ist not resolving, if the user clicked "Allow once".  i can reproduce it with an simulator, if i clicking "Allow once".

**Expected behavior**
`getLocation` should be resolving or thrown a error.

**Tested on:**
 - customer iOS device
 - iOS iPad Air 2
 - iOS Si

**Additional logs**
in Visual Studio Code or in XCode i saw no errors.

**Comment by cmorsucci** [comment]
Same Issue, happens only on IOS

**Comment by jpcarreira** [comment]
I'm also experiencing the exact same thing, function not resolving on the first run. We need to restart the app to go through `getLocation()`

**Comment by Faisalkc4u** [comment]
Was anyone able to solve this ? I am currently facing the same issue. Android works fine its just IOS, even if I restart the app on the simulator it still doesn't ask for permissions. Are there any workarounds for this bug ?

**Comment by ademanuele** [comment]
I have the same issue.

However, I tried creating a simple example for demonstration and could not replicate this.
So perhaps we're doing something wrong?

This is the dummy example I have created:
[location.zip](https://github.com/Lyokone/flutterlocation/files/7092403/location.zip)

**Comment by ademanuele** [comment]
Correction: I can replicate it when giving the simulator a single custom location which doesn't change.

In this case, the first call to `getLocation()` returns, however another call fails.

If you set the simulator's location to "City Run" however, this is not reproducable.
So it seems that this only happens when the location is fixed and unchanging.

**Comment by maerlynflagg** [comment]
> Was anyone able to solve this ? I am currently facing the same issue. Android works fine its just IOS, even if I restart the app on the simulator it still doesn't ask for permissions. Are there any workarounds for this bug ?

the only workaround what i know do, is to timeout the promise, so that the App not freeze and waits for getLocation().

**Comment by LostInDarkMath** [comment]
@Lyokone Could you review the merge request soon please? This is a huge issue for our app. Thanks!

**Comment by martesabt** [comment]
I experienced this only on iOS simulator. Hope the merge request is reviewed soon.

**Comment by LostInDarkMath** [comment]
Maybe requesting the location a second time is valid workaround for this. The first, the `Future` of `getLocation()` does not return, but the second time it does.

So instead of
```dart
LocationData location = await _locator.getLocation();
```
use this:
```dart
    LocationData? location;
    while(location == null) {
      try {
        location = await Future.any<LocationData>([
          _locator.getLocation(), // we hope this returns first
          Future.delayed(const Duration(seconds: 5), () => throw LocatorTimeout()),
        ]);
      } on LocatorTimeout {
        print('LocatorTimeout. Trying again...');
      }
    }
```

Does this work for you?

**Comment by wiseminds** [comment]
> @Lyokone Could you review the merge request soon please? This is a huge issue for our app. Thanks!

I tested out that commit on my app but the issue still persists.

see implementation below

``` location:  
     git:
      url: https://github.com/ademanuele/flutterlocation.git
      path: packages/location
      ref: 6810b7e338173ff26a0a52a25721060a563e625f   ```

**Comment by ademanuele** [comment]
Hi @wiseminds.

 Do you mean you tried the code in this MR and it did not work for you?
Could you describe any change in the behaviour if you notice any please?

**Comment by wiseminds** [comment]
There was no difference.

**Comment by AlpSha** [comment]
> Maybe requesting the location a second time is valid workaround for this. The first, the `Future` of `getLocation()` does not return, but the second time it does.
> 
> So instead of
> 
> ```dart
> LocationData location = await _locator.getLocation();
> ```
> 
> use this:
> 
> ```dart
>     LocationData? location;
>     while(location == null) {
>       try {
>         location = await Future.any<LocationData>([
>           _locator.getLocation(), // we hope this returns first
>           Future.delayed(const Duration(seconds: 5), () => throw LocatorTimeout()),
>         ]);
>       } on LocatorTimeout {
>         print('LocatorTimeout. Trying again...');
>       }
>     }
> ```
> 
> Does this work for you?
Actually the problem is that it doesn't work after first call. So it doesn't matter how many times you try...

**Comment by LostInDarkMath** [comment]
@AlpSha Yes, I think you are right. The problem seems to different on different platforms. I had this problem on bith Android and iOS. On Android, the second call is successful so my workwaround fixes the issue on Android. But on iOS it looks like this is not the case. Did you test on Android or iOS?

**Comment by AlpSha** [comment]
@LostInDarkMath Tested on both. Android works fine without any workaround but iOS has the issue.

**Comment by RedHappyLlama** [comment]
Hi, 

I had the same issue on iOS Simulator. I've just noticed that the issue only occured when the simulator location settings (Features, Location) were set to Custom Location.... Any other setting it seemed to work fine. Might help?

**Comment by ibraah88** [comment]
Hi, 
i have the same issue, do you have any idea how can i fix-it please ?

**Comment by anirudhsharma392** [comment]
Same issue any update ?

**Comment by dvird** [comment]
`
return await _location.getLocation().timeout(Duration(seconds: 5));
`

**Comment by neoacevedo** [comment]
Now in Android the issue is pressent.

**Comment by kashiflab** [comment]
Any update I am getting the issue on Redmi 8.

**Comment by m2sahin** [comment]
The problem persists for iOS.

**Comment by maerlynflagg** [comment]
this week icg noticed that getLocation still doesn't work in the iOS simulator.

the app keeps waiting for a result to come back or at least an error to be thrown.

**Comment by Lyokone** [comment]
Traced this against the current `LocationPlugin.swift`. `getLocation()` appends its result to `pendingLocationResults` before requesting permission (if not yet determined), and the authorization-change handler resolves that queue as soon as status becomes `.authorizedWhenInUse` *or* `.authorizedAlways`:

```swift
guard granted else { return }
...
if !pendingLocationResults.isEmpty || flutterListening {
    clLocationManager?.startUpdatingLocation()
}
```

Core Location reports "Allow Once" as plain `.authorizedWhenInUse` — the same status as "While Using the App" — so there's no code path that treats it differently, and this resolves the pending `getLocation()` call the same way for both choices.

This queuing (and the removal of the old `locationWanted` boolean it replaced) was reworked this year specifically to fix pending `getLocation()` calls being silently dropped around permission-grant timing (#977, #979), which is exactly the scenario this issue describes.

Closing as resolved by the current implementation. If you can still reproduce this on the latest version specifically with "Allow Once", please reopen with repro steps (iOS version + whether this is with a fresh install or a permission already granted from a prior session) — that combination might expose a narrower edge case not covered by the above.
